### PR TITLE
Fix FileHandle Leak in BaseDataReader for MetadataFiles

### PR DESCRIPTION
**Pull request by RussellSpitzer**
The core issue here was that previously when using CloseableIterable.filter
The iterator generated by the wrapped Iterable would be wrapped by a Guava
Iterator regardless of the wrapped iterators closable status. This means
Calling Filter on an iterable would result in a CloseableIterable which
generated incorrectly unclosable Iterators.

To fix this we replace the Guava IterablesFilter operation with the
FilterIterator class. This properly wraps the underlying closable iterator
so that when we generate iterators we still get classes that can close
their wrapped implementations.

### CODE REVIEW

**Comment by RussellSpitzer** [comment]
cc @aokolnychyi + @rdblue

File: api/src/test/java/org/apache/iceberg/io/TestCloseableIterable.java [hunk 1]
```diff
@@ -28,6 +29,36 @@
 
 public class TestCloseableIterable {
 
+  @Test
+  public void testFilterManuallyClosable() throws IOException {
+    TestableCloseableIterable iterable = new TestableCloseableIterable();
+    TestableCloseableIterable.TestableCloseableIterator iterator =
```
Comment (aokolnychyi) [inline api/src/test/java/org/apache/iceberg/io/TestCloseableIterable.java:35]: nit: can we import `TestableCloseableIterator` and fit on one line?

File: api/src/test/java/org/apache/iceberg/io/TestableCloseableIterable.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+
+public class TestableCloseableIterable implements  CloseableIterable<Integer>{
+    private Boolean closed = false;
```
Comment (rdblue) [inline api/src/test/java/org/apache/iceberg/io/TestableCloseableIterable.java:20]: Looks like indentation (2 spaces) is off for this file (4 spaces). Can you fix it?

File: api/src/test/java/org/apache/iceberg/io/TestableCloseableIterable.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.io.IOException;
+
+public class TestableCloseableIterable implements  CloseableIterable<Integer>{
```
Comment (aokolnychyi) [inline api/src/test/java/org/apache/iceberg/io/TestableCloseableIterable.java:19]: Extra space after `implements`

**Comment by aokolnychyi** [comment]
Looks good to me except formatting nits.

File: api/src/test/java/org/apache/iceberg/io/TestableCloseableIterable.java (hunk 2 above)
Comment (RussellSpitzer) [reply api/src/test/java/org/apache/iceberg/io/TestableCloseableIterable.java:20]: Ah looks like "run checkstyle on this project" from IDEA only applies to non test files whoops!

**Comment by RussellSpitzer** [comment]
All cleaned up style wise

**Comment by aokolnychyi** [comment]
I'd put `FilterIterator` into the `io` package instead of `util`. I think it makes more sense there. Other than that, LGTM.

**Comment by aokolnychyi** [comment]
I've tested locally and it solves the problem in metadata tables. I am merging it.

**Pull request merged**